### PR TITLE
Add date handling to axis widgets

### PR DESCRIPTION
**Pull request by Bliss3d**
Same as #2288, but improved now dates get interpolated so its practily speaking infinite

### CODE REVIEW

**Comment by Bliss3d** [comment]
So how are the changes that this and #2287 get merged? Also #2287 need CI approvement :)

**Comment by djhoese** [comment]
It will take time to review these given the size and number of changes (big features). Right now I need to do a bug fix release of vispy first then I can review these and we talk about merging them. But before all that I need to have time to do that review. If any of @vispy/core has spare time, please help me review these two PRs.

**Comment by rougier** [comment]
Can you post some screenshots of the result?

File: vispy/scene/widgets/axis.py [hunk 1]
```diff
@@ -16,16 +16,32 @@ class AxisWidget(Widget):
     ----------
     orientation : str
         Orientation of the axis, 'left' or 'bottom'.
+    axis_mapping : array-like
+        Axis mapping maps a arbitrary array-like object of type numpy array of numpy.datetime64 or
+        list of datetime dates
+        All other types will be converted to strings
+        In case of string mapping the array/list will only be applied to integers
+        (0, 1, 2, 3, 4, 5, etc) not (0.5, 0.666, 0.75, 1.5, etc.) floats won't be displayed,
+        this is not the case if you provide a array of datetime64 or list of datetime dates
```
Comment (brisvag) [inline vispy/scene/widgets/axis.py:25]:
```suggestion
        Axis mapping maps an arbitrary array-like object to the axis ticks. If a list of datetime objects
        or a numpy array of dtype numpy.datetime64, they will be interpolated; otherwise, elements will
        be cast to string and mapped to integer ticks.
```

File: vispy/visuals/axis.py [hunk 2]
```diff
@@ -70,25 +75,50 @@ class AxisVisual(CompoundVisual):
         of 'left', 'center', or 'right', and the second element should be one
         of 'bottom', 'middle', or 'top'. If this is not specified, it is
         determined automatically.
+    axis_mapping : array-like
+        Axis mapping maps an arbitrary array-like object of type numpy array of numpy.datetime64 or
+        list of datetime dates, for better performance use list since numpy will be converted to list anyway
+        All other types will be converted to strings
+        In case of string mapping the array/list will only be applied to integers
+        (0, 1, 2, 3, 4, 5, etc) not (0.5, 0.666, 0.75, 1.5, etc.) floats won't be displayed,
+        this is not the case if you provide a array of datetime64 or list of datetime dates
```
Comment (brisvag) [inline vispy/visuals/axis.py:84]: Same as above, I would reword this.

File: vispy/scene/widgets/axis.py (hunk 1 above)
Comment (brisvag) [reply vispy/scene/widgets/axis.py:25]: I would reword this as above, assuming that I udnerstood what it means.

**Comment by brisvag** [comment]
No worries @Bliss3d, no pressure! Let me know when you want me to review again :)

**Comment by Bliss3d** [comment]
test
So here is my little example
`import numpy as np
import pandas as pd
from vispy import scene, app
from vispy.color import ColorArray

if __name__ == "__main__":

    # vispy.app.use_app("pyglet")

    def update(ev):
        global i
        pos11[:, 1] = np.random.normal(loc=140, scale=5, size=N)
        pos12[:, 1] = np.random.normal(loc=120, scale=5, size=N)
        pos13[:, 1] = np.random.normal(loc=100, scale=5, size=N)
        pos14[:, 1] = np.random.normal(loc=80, scale=5, size=N)
        pos15[:, 1] = np.random.normal(loc=60, scale=5, size=N)

        line11.set_data(pos=pos11, color=ColorArray('yellow'))
        line12.set_data(pos=pos12, color=ColorArray('blue'))
        line13.set_data(pos=pos13, color=(1, 0.1, 0.))
        line14.set_data(pos=pos14, color=(0.25, 0.8, 0.))
        line15.set_data(pos=pos15, color=(1, 1, 1))

    N = 1200

    pos11 = np.zeros((N, 2), dtype=np.float32)
    pos12 = np.zeros((N, 2), dtype=np.float32)
    pos13 = np.zeros((N, 2), dtype=np.float32)
    pos14 = np.zeros((N, 2), dtype=np.float32)
    pos15 = np.zeros((N, 2), dtype=np.float32)


    x_lim = [0., 1200.]
    y_lim1 = [50., 150.]
    y_lim2 = [50., 150.]
    y_lim3 = [50., 150.]
    y_lim4 = [50., 150.]
    y_lim5 = [50., 150.]

    pos11[:, 0] = np.linspace(x_lim[0], x_lim[1], N)
    pos11[:, 1] = np.random.normal(loc=120, size=N)
    pos12[:, 0] = np.linspace(x_lim[0], x_lim[1], N)
    pos12[:, 1] = np.random.normal(loc=100, size=N)
    pos13[:, 0] = np.linspace(x_lim[0], x_lim[1], N)
    pos13[:, 1] = np.random.normal(loc=80, size=N)
    pos14[:, 0] = np.linspace(x_lim[0], x_lim[1], N)
    pos14[:, 1] = np.random.normal(loc=60, size=N)
    pos15[:, 0] = np.linspace(x_lim[0], x_lim[1], N)
    pos15[:, 1] = np.random.normal(loc=140, size=N)


    color = np.ones((N, 4), dtype=np.float32)
    color[:, 0] = np.linspace(0, 1, N)
    color[:, 1] = color[::-1, 0]

    canvas = scene.SceneCanvas(keys='interactive', vsync=True)
    canvas.size = 800, 600
    canvas.show()

    grid = canvas.central_widget.add_grid()
    grid.padding = 10

    vb1 = grid.add_view(row=0, col=1, camera='panzoom')

    x_axis1 = scene.AxisWidget(orientation='bottom', axis_mapping=pd.date_range(start ='1-1-2018', periods = 13, freq="1H").tolist(), date_format_string="%m/%d/%Y, %H:%M:%S")
    x_axis1.stretch = (1, 0.1)
    grid.add_widget(x_axis1, row=1, col=1)
    x_axis1.link_view(vb1)
    y_axis1 = scene.AxisWidget(orientation='left')
    y_axis1.stretch = (0.1, 1)
    grid.add_widget(y_axis1, row=0, col=0)
    y_axis1.link_view(vb1)


    grid_lines1 = scene.visuals.GridLines(parent=vb1.scene)

    line11 = scene.Line(pos11, ColorArray('yellow'), parent=vb1.scene, width=0.8)
    line12 = scene.Line(pos12, ColorArray('blue'), parent=vb1.scene, width=0.8)
    line13 = scene.Line(pos13, (1, 0.1, 0.), parent=vb1.scene, width=0.8)
    line14 = scene.Line(pos14, (0.25, 0.8, 0.), parent=vb1.scene, width=0.8)
    line15 = scene.Line(pos15, (0, 0, 0), parent=vb1.scene, width=0.8)


    vb1.camera.set_range()

    timer = app.Timer()
    timer.connect(update)
    timer.start(0)

    canvas.measure_fps()
    app.run()`

Those insanely long 000000xxx date strings a coming from pandas because its a pandas timestamp and I wasnt sure how to parse them without importing pandas. 
the date interpolation rounding does have some issues, but in sensable ranges it works quite well - Its limited to go down to secounds after that it goes crazy. In case people want to display ms/µs/ns they would need to disable the date interpolation.
This maybe could be improved and split into two seperate function one for interpolation between ticks and one form just extending the dates along the axis linearly the current implementation does both. 

The next gif is done with numpy arrange datetime 
x_axis1 = scene.AxisWidget(orientation='bottom', axis_mapping=np.arange('2000-01-01', '2000-01-02',np.timedelta64(1,'h'), dtype='datetime64'), date_format_string="%m/%d/%Y, %H:%M:%S")

test2

Some feedback would be great if someone is better than me with time and time rounding please feel free to improve it :)

**Comment by brisvag** [comment]
> In case people want to display ms/µs/ns they would need to disable the date interpolation. This maybe could be improved and split into two seperate function one for interpolation between ticks and one form just extending the dates along the axis linearly the current implementation does both.

I think it's fine to have to disable date interpolation for ms resolution; I don't think we need to hold up this PR for that, as this is already a net improvement over what we have. The changes you propose can easily come up in a followup PR, in my opinion.

**Comment by Bliss3d** [comment]
In the next couple of days i'll add some bug fixes and improvements, I found some issues and hope to iron them out

**Comment by Bliss3d** [comment]
Repleced by #2402
